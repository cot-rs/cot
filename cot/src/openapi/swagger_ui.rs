//! The OpenAPI (Swagger) UI router.
//!
//! This module provides a [`cot::App`] which serves the OpenAPI (Swagger) UI.

use std::borrow::Cow;
use std::sync::Arc;

use bytes::Bytes;
use cot::request::Request;
use swagger_ui_redist::SwaggerUiStaticFile;

use crate::request::RequestExt;
use crate::response::{Response, ResponseExt};
use crate::router::{Route, Router};
use crate::{App, Body, StatusCode};

/// A wrapper around the Swagger UI functionality.
///
/// This struct serves the Swagger UI interface for OpenAPI documentation.
/// It can be registered with a Cot application to provide interactive API
/// documentation at a specified URL path.
///
/// # Example
///
/// ```
/// use cot::openapi::swagger_ui::SwaggerUi;
/// use cot::project::{AppBuilder, RegisterAppsContext};
///
/// fn register_apps(apps: &mut AppBuilder, _context: &RegisterAppsContext) {
///     // Register SwaggerUI at the "/swagger" path
///     apps.register_with_views(SwaggerUi::new(), "/swagger");
/// }
/// ```
#[derive(Debug, Clone)]
pub struct SwaggerUi {
    inner: swagger_ui_redist::SwaggerUi,
    serve_openapi: bool,
}

async fn openapi_json(request: Request) -> cot::Result<Response> {
    let openapi = aide::openapi::OpenApi {
        paths: Some(request.router().as_api()),
        ..Default::default()
    };

    Response::new_json(StatusCode::OK, &openapi)
}

impl Default for SwaggerUi {
    fn default() -> Self {
        Self::new()
    }
}

impl SwaggerUi {
    /// Creates a new [`SwaggerUi`] that serves the OpenAPI JSON at the default
    /// path.
    #[must_use]
    pub fn new() -> Self {
        Self::new_impl(Cow::Borrowed("openapi.json"), true)
    }

    /// Creates a new [`SwaggerUi`] that serves the OpenAPI JSON at the
    /// specified path.
    #[must_use]
    pub fn with_api_at<P: Into<Cow<'static, str>>>(openapi_path: P) -> Self {
        Self::new_impl(openapi_path.into(), false)
    }

    fn new_impl(openapi_path: Cow<'static, str>, serve_openapi: bool) -> Self {
        let mut swagger_ui = swagger_ui_redist::SwaggerUi::new();
        swagger_ui.config().urls([openapi_path]);
        for static_file in SwaggerUiStaticFile::all() {
            let file_path = format!("/static/{}", Self::static_file_path(*static_file));
            swagger_ui.override_file_path(*static_file, file_path);
        }

        Self {
            inner: swagger_ui,
            serve_openapi,
        }
    }

    fn static_file_path(static_file: SwaggerUiStaticFile) -> String {
        format!("swagger/{}", static_file.file_name())
    }
}

impl App for SwaggerUi {
    fn name(&self) -> &'static str {
        "swagger-ui"
    }

    fn router(&self) -> Router {
        let swagger_ui = Arc::new(self.inner.clone());
        let swagger_handler = async move || {
            Ok(Response::new_html(
                StatusCode::OK,
                Body::fixed(swagger_ui.serve().map_err(cot::Error::custom)?),
            ))
        };

        let mut urls = vec![Route::with_handler("/", swagger_handler)];
        if self.serve_openapi {
            urls.push(Route::with_handler("/openapi.json", openapi_json));
        }
        Router::with_urls(urls)
    }

    fn static_files(&self) -> Vec<(String, Bytes)> {
        swagger_ui_redist::SwaggerUi::static_files()
            .iter()
            .map(|(static_file_id, data)| {
                let path = Self::static_file_path(*static_file_id);
                let bytes = Bytes::from_static(data);
                (path, bytes)
            })
            .collect()
    }
}
