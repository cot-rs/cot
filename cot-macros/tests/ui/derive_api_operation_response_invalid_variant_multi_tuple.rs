use cot::openapi::ApiOperationResponse;

#[derive(ApiOperationResponse)]
enum MyResponse {
    A(u32, String),
}

fn main() {}
