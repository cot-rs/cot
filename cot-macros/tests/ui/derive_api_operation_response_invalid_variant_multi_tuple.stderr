error: only tuple variants with a single field are supported
 --> tests/ui/derive_api_operation_response_invalid_variant_multi_tuple.rs:3:10
  |
3 | #[derive(ApiOperationResponse)]
  |          ^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the derive macro `ApiOperationResponse` (in Nightly builds, run with -Z macro-backtrace for more info)
